examples/api/lib/material/autocomplete/autocomplete.5.dart

```diff
@@ -0,0 +1,249 @@
+  static Future<Iterable<String>> search(String query) async {
+    await Future<void>.delayed(fakeAPIDuration); // Fake 1 second delay.
+    if (query == '') {
+      return const Iterable<String>.empty();
+    }
+    return _kOptions.where((String option) {
+      return option.contains(query.toLowerCase());
+    });
+  }
+}
+
+typedef _Debounceable<S, T> = Future<S?> Function(T parameter);
+
+/// Returns a new function that is a debounced version of the given function.
+///
+/// This means that the original function will be called only after no calls
+/// have been made for the given Duration.
+_Debounceable<S, T> _debounce<S, T>(_Debounceable<S?, T> function) {
+  _DebounceTimer? debounceTimer;
+
+  return (T parameter) async {
+    if (debounceTimer != null && !debounceTimer!.isCompleted) {
+      debounceTimer!.cancel();
+    }
+    debounceTimer = _DebounceTimer();
+    try {
+      await debounceTimer!.future;
+    } catch (error) {
+      if (error is _CancelException) {
+        return null;
+      }
+      rethrow;
+    }
+    return function(parameter);
+  };
+}
+
+// A wrapper around Timer used for debouncing.
+class _DebounceTimer {
+  _DebounceTimer() {
+    _timer = Timer(debounceDuration, _onComplete);
+  }
+
+  late final Timer _timer;
+  final Completer<void> _completer = Completer<void>();
+
+  void _onComplete() {
+    _completer.complete();
+  }
+
+  Future<void> get future => _completer.future;
+
+  bool get isCompleted => _completer.isCompleted;
+
+  void cancel() {
+    _timer.cancel();
+    _completer.completeError(const _CancelException());
+  }
+}
+
+// An exception indicating that the timer was canceled.
+class _CancelException implements Exception {
+  const _CancelException();
+}
+
+class AutocompleteOverlay extends StatelessWidget {
+  const AutocompleteOverlay({
+    super.key,
+    required this.onSelected,
+    required this.options,
+  });
+  final AutocompleteOnSelected<String> onSelected;
+  final Iterable<String> options;
+
+  @override
+  Widget build(BuildContext context) {
+    return Align(
+      alignment: Alignment.topLeft,
+      child: Material(
+        elevation: 4.0,
+        child: ConstrainedBox(
+          constraints: const BoxConstraints(maxHeight: 200, maxWidth: 500),
+          child: ListView.builder(
+            padding: EdgeInsets.zero,
+            shrinkWrap: true,
+            itemCount: options.length,
+            itemBuilder: (BuildContext context, int index) {
+              final String option = options.elementAt(index);
+              return InkWell(
+                onTap: () {
+                  onSelected(option);
+                },
+                child: Container(
+                  color: Theme.of(context).focusColor,
+                  padding: const EdgeInsets.all(16.0),
+                  child: Text(option),
+                ),
+              );
+            },
+          ),
+        ),
+      ),
+    );
+  }
+}
```

Comment:
One small bug report: Type something and then quickly delete it, so the field is empty. Expected: it immediately says "type something". Acutal: it still says "loading" for awhile first.

Reply:
Also another bug report: Type something so that no options are found. Type an extra character. Expected: It still shows 'no options found'. Actual: It still says 'loading' for awhile first.

Reply:
> One small bug report: Type something and then quickly delete it, so the field is empty. Expected: it immediately says "type something". Acutal: it still says "loading" for awhile first.

Fixed by adding a variable `isFieldEmpty`, then only returning a loading view when this variable is false.

> Also another bug report: Type something so that no options are found. Type an extra character. Expected: It still shows 'no options found'. Actual: It still says 'loading' for awhile first.

Fixed by stopping once no options are found if an extra character (or more) is entered.